### ISSUE TITLE
kubectl apply --prune and kubectl diff --prune have opposite behavior

### ISSUE THREAD

**Issue by muawiakh** [kubernetes/kubectl#1355]
**What happened**:

Knowing that `--prune` is in alpha state but I think there is a discrepancy between: 

```bash
kubectl apply --prune --selector=<some labels> -k /path/to/kustomize/env/dir
OR 
kubectl apply --prune --selector=<some labels> --dry-run='server' -k /path/to/kustomize/env/dir
```

**AND** when we try to do a diff, the objects to be pruned are not correct e.g.

```bash
kubectl diff --prune --selector=<some labels> -k -/path/to/kustomize/env/dir
```

I think the issue is with the `--selector` flag in the `kubectl diff` command. With the selector enabled, `diff` shows everything else as well that is not part of the configuration being applied even though it should only be concerned with the the labels we have provided. `apply` and `apply --dry-run='server'`  display the correct resources that will or will not be pruned AND if we skip the `--selector` from the `diff` command it shows the correct configuration changes.

**How to reproduce it (as minimally and precisely as possible)**:
Using this `manifest.yaml`

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    kubernetes.io/issue: reproduce
    app: nginx
spec:
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: nginx
        ports:
        - containerPort: 8080
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: postgres
spec:
  selector:
    matchLabels:
      app: postgres
  template:
    metadata:
      labels:
        app: postgres
    spec:
      containers:
        - name: postgres
          image: postgres
          ports:
            - containerPort: 5432
          env:
            - name: POSTGRES_DB
              value: mydatabase
            - name: POSTGRES_USER
              value: myuser
            - name: POSTGRES_PASSWORD
              value: mypassword
```
Apply the above configuration:

```bash
kubectl apply -f manifest.yaml
```
Changing the image version of the `nginx` container with the diff:

```bash
<       - image: nginx
---
>       - image: nginx:newimage
```

```bash
# Correct behavior because it should only concerned with the label I mentioned
$ kubectl apply --prune --selector=kubernetes.io/issue=reproduce --dry-run='server' -f sampledeployment.yaml

deployment.apps/nginx configured (server dry run)
```

```bash
# Not sure if this is correct, because it will prune all objects 
# and we didn't mention any selector so it should include postgres as well

$ kubectl diff --prune  -f sampledeployment.yaml
diff -N - /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/LIVE-1648926106/apps.v1.Deployment.default.nginx /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/MERGED-2140770436/apps.v1.Deployment.default.nginx
--- /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/LIVE-1648926106/apps.v1.Deployment.default.nginx	2023-01-19 14:07:02
+++ /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/MERGED-2140770436/apps.v1.Deployment.default.nginx	2023-01-19 14:07:02
@@ -6,7 +6,7 @@
     kubectl.kubernetes.io/last-applied-configuration: |
       {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app":"nginx","kubernetes.io/issue":"reproduce"},"name":"nginx","namespace":"default"},"spec":{"selector":{"matchLabels":{"app":"nginx"}},"template":{"metadata":{"labels":{"app":"nginx"}},"spec":{"containers":[{"image":"nginx","name":"nginx","ports":[{"containerPort":8080}]}]}}}}
   creationTimestamp: "2023-01-19T13:04:09Z"
-  generation: 1
+  generation: 2
   labels:
     app: nginx
     kubernetes.io/issue: reproduce
@@ -33,7 +33,7 @@
         app: nginx
     spec:
       containers:
apiVersion: apps/v1
-      - image: nginx
+      - image: nginx:newimage
         imagePullPolicy: Always
         name: nginx
         ports
```

Here is the behavior of `apply` when we use `--prune --all`
```bash
$ kubectl apply --prune --all  --dry-run='server' -f sampledeployment.yaml
deployment.apps/nginx unchanged (server dry run)
deployment.apps/postgres unchanged (server dry run)
```

```bash
#  In my understanding the opposite because we specified the label only for the
# nginx deployment but it shows the difference with postgres as well

$ kubectl diff --prune --selector=kubernetes.io/issue=reproduce -f sampledeployment.yaml
diff -N - /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/LIVE-2073164283/apps.v1.Deployment.default.nginx /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/MERGED-3879221752/apps.v1.Deployment.default.nginx
--- /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/LIVE-2073164283/apps.v1.Deployment.default.nginx	2023-01-19 14:06:29
+++ /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/MERGED-3879221752/apps.v1.Deployment.default.nginx	2023-01-19 14:06:29
@@ -6,7 +6,7 @@
     kubectl.kubernetes.io/last-applied-configuration: |
       {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app":"nginx","kubernetes.io/issue":"reproduce"},"name":"nginx","namespace":"default"},"spec":{"selector":{"matchLabels":{"app":"nginx"}},"template":{"metadata":{"labels":{"app":"nginx"}},"spec":{"containers":[{"image":"nginx","name":"nginx","ports":[{"containerPort":8080}]}]}}}}
   creationTimestamp: "2023-01-19T13:04:09Z"
-  generation: 1
+  generation: 2
   labels:
     app: nginx
     kubernetes.io/issue: reproduce
@@ -33,7 +33,7 @@
         app: nginx
     spec:
       containers:
-      - image: nginx
+      - image: nginx:newimage
         imagePullPolicy: Always
         name: nginx
         ports:
diff -N - /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/LIVE-2073164283/apps.v1.Deployment.default.postgres /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/MERGED-3879221752/apps.v1.Deployment.default.postgres
--- /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/LIVE-2073164283/apps.v1.Deployment.default.postgres	2023-01-19 14:06:30
+++ /var/folders/fr/_gsj7b8x66524j_n2bktfgsr0000gn/T/MERGED-3879221752/apps.v1.Deployment.default.postgres	1970-01-01 01:00:00
@@ -1,168 +0,0 @@
-apiVersion: apps/v1
-kind: Deployment
-metadata:
-  annotations:
-    deployment.kubernetes.io/revision: "1"
-    kubectl.kubernetes.io/last-applied-configuration: |
-      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"name":"postgres","namespace":"default"},"spec":{"selector":{"matchLabels":{"app":"postgres"}},"template":{"metadata":{"labels":{"app":"postgres"}},"spec":{"containers":[{"env":[{"name":"POSTGRES_DB","value":"mydatabase"},{"name":"POSTGRES_USER","value":"myuser"},{"name":"POSTGRES_PASSWORD","value":"mypassword"}],"image":"postgres","name":"postgres","ports":[{"containerPort":5432}]}]}}}}
-  creationTimestamp: "2023-01-19T13:04:09Z"
-  generation: 1
-  managedFields:
-  - apiVersion: apps/v1
-    fieldsType: FieldsV1
-    fieldsV1:
-      f:metadata:
-        f:annotations:
-          .: {}
-          f:kubectl.kubernetes.io/last-applied-configuration: {}
-      f:spec:
-        f:progressDeadlineSeconds: {}
-        f:replicas: {}
-        f:revisionHistoryLimit: {}
-        f:selector: {}
-        f:strategy:
-          f:rollingUpdate:
-            .: {}
-            f:maxSurge: {}
-            f:maxUnavailable: {}
-          f:type: {}
-        f:template:
-          f:metadata:
-            f:labels:
-              .: {}
-              f:app: {}
-          f:spec:
-            f:containers:
-              k:{"name":"postgres"}:
-                .: {}
-                f:env:
-                  .: {}
-                  k:{"name":"POSTGRES_DB"}:
-                    .: {}
-                    f:name: {}
-                    f:value: {}
-                  k:{"name":"POSTGRES_PASSWORD"}:
-                    .: {}
-                    f:name: {}
-                    f:value: {}
-                  k:{"name":"POSTGRES_USER"}:
-                    .: {}
-                    f:name: {}
-                    f:value: {}
-                f:image: {}
-                f:imagePullPolicy: {}
-                f:name: {}
-                f:ports:
-                  .: {}
-                  k:{"containerPort":5432,"protocol":"TCP"}:
-                    .: {}
-                    f:containerPort: {}
-                    f:protocol: {}
-                f:resources: {}
-                f:terminationMessagePath: {}
-                f:terminationMessagePolicy: {}
-            f:dnsPolicy: {}
-            f:restartPolicy: {}
-            f:schedulerName: {}
-            f:securityContext: {}
-            f:terminationGracePeriodSeconds: {}
-    manager: kubectl-client-side-apply
-    operation: Update
-    time: "2023-01-19T13:04:09Z"
-  - apiVersion: apps/v1
-    fieldsType: FieldsV1
-    fieldsV1:
-      f:metadata:
-        f:annotations:
-          f:deployment.kubernetes.io/revision: {}
-      f:status:
-        f:availableReplicas: {}
-        f:conditions:
-          .: {}
-          k:{"type":"Available"}:
-            .: {}
-            f:lastTransitionTime: {}
-            f:lastUpdateTime: {}
-            f:message: {}
-            f:reason: {}
-            f:status: {}
-            f:type: {}
-          k:{"type":"Progressing"}:
-            .: {}
-            f:lastTransitionTime: {}
-            f:lastUpdateTime: {}
-            f:message: {}
-            f:reason: {}
-            f:status: {}
-            f:type: {}
-        f:observedGeneration: {}
-        f:readyReplicas: {}
-        f:replicas: {}
-        f:updatedReplicas: {}
-    manager: kube-controller-manager
-    operation: Update
-    subresource: status
-    time: "2023-01-19T13:04:11Z"
-  name: postgres
-  namespace: default
-  resourceVersion: "103966"
-  uid: 42a30729-135b-44b7-9573-8a2e9ab6f922
-spec:
-  progressDeadlineSeconds: 600
-  replicas: 1
-  revisionHistoryLimit: 10
-  selector:
-    matchLabels:
-      app: postgres
-  strategy:
-    rollingUpdate:
-      maxSurge: 25%
-      maxUnavailable: 25%
-    type: RollingUpdate
-  template:
-    metadata:
-      creationTimestamp: null
-      labels:
-        app: postgres
-    spec:
-      containers:
-      - env:
-        - name: POSTGRES_DB
-          value: mydatabase
-        - name: POSTGRES_USER
-          value: myuser
-        - name: POSTGRES_PASSWORD
-          value: mypassword
-        image: postgres
-        imagePullPolicy: Always
-        name: postgres
-        ports:
-        - containerPort: 5432
-          protocol: TCP
-        resources: {}
-        terminationMessagePath: /dev/termination-log
-        terminationMessagePolicy: File
-      dnsPolicy: ClusterFirst
-      restartPolicy: Always
-      schedulerName: default-scheduler
-      securityContext: {}
-      terminationGracePeriodSeconds: 30
-status:
-  availableReplicas: 1
-  conditions:
-  - lastTransitionTime: "2023-01-19T13:04:11Z"
-    lastUpdateTime: "2023-01-19T13:04:11Z"
-    message: Deployment has minimum availability.
-    reason: MinimumReplicasAvailable
-    status: "True"
-    type: Available
-  - lastTransitionTime: "2023-01-19T13:04:09Z"
-    lastUpdateTime: "2023-01-19T13:04:11Z"
-    message: ReplicaSet "postgres-7b47d5f7b4" has successfully progressed.
-    reason: NewReplicaSetAvailable
-    status: "True"
-    type: Progressing
-  observedGeneration: 1
-  readyReplicas: 1
-  replicas: 1
-  updatedReplicas: 1
```

**What you expected to happen**:
In my opinion the behaviour with `--selector` and without should be the opposite. 

**Anything else we need to know?**:

```
kubectl version --short
Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
Client Version: v1.25.4
Kustomize Version: v4.5.7
Server Version: v1.25.3
```

Could be reproduced anywhere we have `--prune` flag available with `kubectl diff`.

**Comment by ardaguclu** [comment]
Thanks for raising issue @muawiakh.  Previous versions, `selector` was not tied to `diff` command and it was fixed with https://github.com/kubernetes/kubernetes/pull/114863 in 1.27. Is it possible to check with the latest version, if it works the way you expect?

**Comment by muawiakh** [comment]
I just checked with `v1.26.0` as well and it was the same issue.

**Comment by brianpursley** [comment]
This is fixed in 1.27 https://github.com/kubernetes/kubernetes/pull/114863
and is planned to be backported to 1.24

/close
